Introduce C++ example

Lorak-mmk opened the pull request:
PR introduces a care-pet example for our cpp-rs driver (but should of course work just as fine with datastax cpp-driver or our legacy cpp-driver).
Example is written in C++. Since managing dependencies in C++ is a pain, I only used Boost libraries.
The example has the same functionality as others. It produces a binary, that has 3 modes:
- Migrator that creates schema
- Sensor that creates Owner, Pet, two Sensor, and produces some data for those sensors
- Server, that servs pseudo-REST API allowing user to query the data. The API is the same as with other examples.

Disclaimer: An AI agent was used for initial development. All code written by it was reviewed by me, and it most cases refactored / rewritten.

To do / consider: server uses synchronous API of the driver. It is absolutely not a good practice, but changing the code to use async would be quite difficult and time consuming. Do we need it for such example?

dkropachev commented on cpp/docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+        image: scylladb/scylla
+        container_name: carepet-scylla1
+        command: --smp 1
+        ports:
+            - "9042:9042"
```
don't need it

dkropachev commented on cpp/src/common/database.cpp:
```diff
@@ -0,0 +1,198 @@
+#include <chrono>
```
Just observation, it looks like something like serialization code from this file has to be part of the driver

dkropachev commented on cpp/src/main.cpp:
```diff
@@ -0,0 +1,58 @@
+
+int main(int argc, char* argv[]) {
+    po::options_description desc("Allowed options");
+    // clang-format off
```
why ?

dkropachev replied:
I get it, but TBH, I am not fan of having formatter and then disable it for part of the code just because you don't like results, you either have it or don't.
Again, just observation, not asking to fix it.

Lorak-mmk replied:
Maybe there is some clang-format config to fix it, idk.

dkropachev commented:
Can we have a cicd for this project with linting, building and testing, to make sure it actually works ?

dkropachev commented:
> > Can we have a cicd for this project with linting, building and testing, to make sure it actually works ?
> 
> It would be great, but I don't see how this should be part of this PR.

Why not ?

Lorak-mmk commented:
> Why not ?

PRs should do one thing. Lack of CI/CD is not specific to the C++ example that I introduce here - it is a general problem in this repo. Thus, it is a separate task.

dkropachev commented:
> > Why not ?
> 
> PRs should do one thing. 

It should, I don't see how having your code tested in a separate task, aren't we require any new functionality have unit and integration test coverage ? don't see how it is different ?

> Lack of CI/CD is not specific to the C++ example that I introduce here - it is a general problem in this repo. Thus, it is a separate task.

Completely disagree, I am struggle to see how it justifies not having your code tested, you can do that complete isolated, to be triggered only on changes in this particular project.

More over, now to check if it works I need to check it out, compile, spin up docker, and go through some validation steps.
And not only me, but anyone who wants to review it.

Lorak-mmk commented on cpp/src/common/database.cpp:
```diff
@@ -0,0 +1,182 @@
+}
+
+const char* cass_value_type_name(CassValueType type) {
+#define CASS_VALUE_TYPE_CASE(name, type, cql, klass)                           \
+    case name:                                                                 \
+        return cql;
+    switch (type) {
+        CASS_VALUE_TYPE_MAPPING(CASS_VALUE_TYPE_CASE);
```
This is a common pattern when defining enums in C. You could just define an enum, and manually specify its variants, and be done with it.
But then you have no way to attach any data to the enum variants (like stringified variant name), and users don't have easy way to write certain stuff.
The solution is to define mapping. Mapping is a macro, that takes another macro X as an arguments, and calls X for each enum variant, passing as argument the variant name, and possibly other stuff.
In this specific case, arguments to X are variant name, cql type id, cql stringified name, and java class name). Definition looks like this:
```c
#define CASS_VALUE_TYPE_MAPPING(XX) \
  XX(CASS_VALUE_TYPE_CUSTOM,  0x0000, "", "") \
  XX(CASS_VALUE_TYPE_ASCII,  0x0001, "ascii", "org.apache.cassandra.db.marshal.AsciiType") \
  XX(CASS_VALUE_TYPE_BIGINT,  0x0002, "bigint", "org.apache.cassandra.db.marshal.LongType") \
  XX(CASS_VALUE_TYPE_BLOB,  0x0003, "blob", "org.apache.cassandra.db.marshal.BytesType") \
...
```

Then the `CASS_VALUE_TYPE` enum is defined using this mapping macro:
```c
typedef enum CassValueType_ {
  CASS_VALUE_TYPE_UNKNOWN = 0xFFFF,
#define XX_VALUE_TYPE(name, type, cql, klass) name = type,
  CASS_VALUE_TYPE_MAPPING(XX_VALUE_TYPE)
#undef XX_VALUE_TYPE
  /* @cond IGNORE */
  CASS_VALUE_TYPE_LAST_ENTRY
  /* @endcond */
} CassValueType;
```

In enum definition, the helper macro (XX_VALUE_TYPE) produces a line of enum definition. In this PR, `CASS_VALUE_TYPE_CASE` produces one case of switch-case statement.

Lorak-mmk commented:
- Fixed "uknown" typo
- Added missing `static`s
- Fixed vertical indentation between methods
- Added vertical separation around helper macro.
- Removed port from docker-compose.yml
- Reworded commit message.

Lorak-mmk commented:
> aren't we require any new functionality have unit and integration test coverage ?

Who's "we" here? Maintainers of this repository definitely don't require this:
- In August, C# example added by Sylwia was merged, without tests / CI
- In May, Rust example was totally refactored, but no tests and no CI were added

So If "we" is to be understood as "maintainers of this repository" or "people working on this repository" then the response is a definite no - test coverage is not required, and there is no precedence here for requiring it.

> Completely disagree, I am struggle to see how it justifies not having your code tested, you can do that complete isolated, to be triggered only on changes in this particular project.

What is the difference between C++ example and other examples? Why is this one the only one that must be tested?
My intent in this PR was to add a C++ example, based on other examples and as similar to other examples as possible. Adding a CI to this repo goes beyond the scope of this.

dkropachev commented:
> > aren't we require any new functionality have unit and integration test coverage ?
> 
> Who's "we" here? Maintainers of this repository definitely don't require this:

They don't, but code we wrote is our responsibility, if it get's to update it we get to do that, if it get's broken we get to fix it.
Probably they could have fixed it in some simple case, but, if there is no CICD they can't event tell if it is broken or not.

> * In August, C# example added by Sylwia was merged, without tests / CI
> * In May, Rust example was totally refactored, but no tests and no CI were added

It was mistake and I will fix it.
 
> So If "we" is to be understood as "maintainers of this repository" or "people working on this repository" then the response is a definite no - test coverage is not required, and there is no precedence here for requiring it.

> > Completely disagree, I am struggle to see how it justifies not having your code tested, you can do that complete isolated, to be triggered only on changes in this particular project.
> 
> What is the difference between C++ example and other examples? Why is this one the only one that must be tested in CI? My intent in this PR was to add a C++ example, based on other examples and as similar to other examples as possible. Adding a CI to this repo goes beyond the scope of this.

It was mistake from the begging, it doesn't mean that we get to continue this bad trend.
Also I want to point out that since you are working on it, compiling it, testing, it is pretty easy for you to come up with the workflow, you basically already have all the commands in our terminal history.

Whereas if it is done by someone else it will be way harder for them to do that.

If some part if it, say integration test requires significant effort, let's start with formating, linting, compiling and unit testing.

tzach commented:
The docs looks good.
I try to compile on Ubuntu and failed:
```
cpp/build$ cmake ..
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.30 or higher is required.  You are running version 3.28.3

``` 

```
cmake is already the newest version (3.28.3-1build7)
```

Not a showstopper, but do we need the latest cmake?
